specification/dartLangSpec.tex

```diff
@@ -14793,71 +14794,151 @@ \subsubsection{For-in}
+
+\LMHash{}%
+Otherwise, $D$ is not empty:
+Let $V$ be the declared type of \id.
+\commentary{%
+Type inference is assumed to have taken place already
+(\ref{overview}),
+so $V$ may be present explicitly or it may have been inferred.%
+}
+A compile-time error occurs unless $T_{\metavar{elm}}$ may be assigned to $V$.
+
+\LMHash{}%
+In either case
+(\commentary{with an empty or non-empty $D$}),
+a fresh scope, the
+\IndexCustom{scope of the asynchronous for-in statement $s$}{%
+  scope!asynchronous for-in statement},
+is introduced which has the enclosing scope of $s$ as its enclosing scope,
+and which is the current scope for the body $s_b$.
+If $D$ is non-empty it introduces a local variable into
+the scope of $s$ with the name \id{} and the declared type $V$.
 
 \LMHash{}%
 Let $D$ be derived from \syntax{<finalConstVarOrType>?}.
-Execution of a for-in statement, $f$, of the form
-\code{\AWAIT{} \FOR{} ($D$ \id{} \IN{} $e$) $s$}
+Execution of a for-in statement, $s$, of the form
+`\code{\AWAIT\,\,\FOR\,\,($D$\,\,\id\,\,\IN\,\,$e$)\,\,$s_b$}'
 proceeds as follows:
 
 \LMHash{}%
 The expression $e$ is evaluated to an object $o$.
 % This error can occur due to implicit casts and null.
-It is a dynamic type error if $o$ is not an instance of a class that implements \code{Stream}.
-It is a compile-time error if $D$ is empty and \id{} is a final variable.
+It is a dynamic type error if $o$ is not an instance of
+a class that implements \code{Stream}.
 
 \LMHash{}%
-The stream associated with the innermost enclosing asynchronous for loop, if any, is paused.
+The stream associated with the innermost enclosing asynchronous for loop,
+if any, is paused.
 The stream $o$ is listened to, producing a stream subscription $u$,
 and execution of the asynchronous for-in loop is suspended
 until a stream event is available.
-\commentary{
-This allows other asynchronous events to execute while this loop is waiting for stream events.
+\commentary{%
+This allows other asynchronous events to execute
+while this loop is waiting for stream events.%
 }
 
-Pausing an asynchronous for loop means pausing the associated stream subscription.
+Pausing an asynchronous for loop means
+pausing the associated stream subscription.
 A stream subscription is paused by calling its \code{pause} method.
-If the subscription is already paused, an implementation may omit further calls to \code{pause}.
+If the subscription is already paused,
+an implementation may omit further calls to \code{pause}.
 
-\commentary{
-The \code{pause} call can throw, although that should never happen for a correctly implemented stream.
+\commentary{%
+The \code{pause} call can throw,
+although that should never happen for a correctly implemented stream.%
 }
 
 \LMHash{}%
 For each \Index{data event} from $u$,
-the statement $s$ is executed with \id{} bound to the value of the current data event.
+the statement $s_b$ is executed with \id{} bound to
+the value of the current data event.
```

Comment:
That's actually imprecise, going on wrong. There is a *new* variable declared for each iteration, and we need to say that. Somehow.

So, it's more like we evaluate `$D$ \id = $v$` where `$v$` is a fresh variable bound to the value of the current data event in a scope surrounding the one used for `$s_b$`.

So, in general, I guess we can rewrite `await for (D id in e) s` to `await for (tmp in e) { D id = tmp; { s }}`, where `tmp` is a fresh local variable of type `$T_elm$`.

Reply:
Now specifying in the dynamic semantics that each iteration uses a fresh variable. This differs from the approach used with a synchronous for-in because it's semantics rather than syntactic (desugaring), but we do introduce the notion of fresh variables in order to specify the semantics of a block statement and such, so it's not a new concept in the spec, just a difference that stands out because for-in is desugared and asynchronous for-in is not. So I believe we can defend both forms, and the discrepancy is a viable choice (even though it would have been a more homogeneous result if we could desugar both forms, or if we would write out the semantics of both forms and avoid the desugaring).